OrbitService/ProducerSideServiceImpl.cpp

```diff
@@ -0,0 +1,302 @@
+        service_state_mutex_.AwaitWithTimeout(absl::Condition(
+                                                  +[](ServiceState* service_state) {
+                                                    return service_state->exit_requested ||
+                                                           service_state->capture_status !=
+                                                               CaptureStatus::kCaptureStarted;
+                                                  },
+                                                  &service_state_),
+                                              kCheckExitSendCommandsThreadInterval);
+        service_state_mutex_.Unlock();
+      } break;
+
+      case CaptureStatus::kCaptureStopping: {
+        service_state_mutex_.AwaitWithTimeout(absl::Condition(
+                                                  +[](ServiceState* service_state) {
+                                                    return service_state->exit_requested ||
+                                                           service_state->capture_status !=
+                                                               CaptureStatus::kCaptureStopping;
+                                                  },
+                                                  &service_state_),
+                                              kCheckExitSendCommandsThreadInterval);
+        service_state_mutex_.Unlock();
+      } break;
+
+      case CaptureStatus::kCaptureFinished: {
+        service_state_mutex_.AwaitWithTimeout(absl::Condition(
+                                                  +[](ServiceState* service_state) {
+                                                    return service_state->exit_requested ||
+                                                           service_state->capture_status !=
+                                                               CaptureStatus::kCaptureFinished;
+                                                  },
+                                                  &service_state_),
+                                              kCheckExitSendCommandsThreadInterval);
+        service_state_mutex_.Unlock();
+      } break;
+    }
+  }
+}
+
+void ProducerSideServiceImpl::ReceiveEventsThread(
+    grpc::ServerContext* /*context*/,
+    grpc::ServerReaderWriter<orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse,
+                             orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest>* stream,
+    bool* all_events_sent_received) {
+  orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest request;
+  while (stream->Read(&request)) {
+    {
+      absl::MutexLock lock{&service_state_mutex_};
+      if (service_state_.exit_requested) {
+        break;
+      }
+    }
+
+    switch (request.event_case()) {
+      case orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest::kBufferedCaptureEvents: {
+        absl::MutexLock lock{&capture_event_buffer_mutex_};
+        if (capture_event_buffer_ != nullptr) {
```

Comment:
In what cases would the buffer be nullptr? Is that reasonable? Should this be rather a check, or do we want to log some error in this case?

Reply:
Okay, thanks for explanation. I leave it up to you on whether to log errors here or not.